File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList();
        generateParanthesisRecursion("",n,n,0,result);
        return result;
    }

    public void generateParanthesisRecursion(String brackets, int open, int closed, int balance, List<String> result) {
        if(open == 0 && closed == 0) {
            result.add(brackets);
            return;
        }
        if(balance == 0) {
            generateParanthesisRecursion(brackets + "(",open-1,closed,balance + 1,result);
        } else {
            if(open != 0 && closed != 0) {
                generateParanthesisRecursion(brackets + "(",open - 1,closed,balance + 1,result);
                generateParanthesisRecursion(brackets + ")",open,closed-1,balance-1,result);
            } else if(open == 0) {
                generateParanthesisRecursion(brackets + ")",open,closed-1,balance-1,result);
            }
        }
    }
}

File: 22-generate-parentheses/generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        result = []
        open,closed,balance = n,n,0
        s = ""
        self.recursive_generation(s,open,closed,balance,result)
        return result
    
    def recursive_generation(self, s, open, closed, balance,result):
        if open == 0 and closed == 0:
            result.append(s)
            return
        
        if open == 0:
            self.recursive_generation(s + ")",open,closed - 1,balance - 1,result)
        elif balance == 0:
            self.recursive_generation(s + "(",open - 1,closed,balance + 1,result)
        else:
            self.recursive_generation(s + "(",open - 1,closed,balance + 1,result)
            self.recursive_generation(s + ")",open,closed -1,balance - 1,result)

        return 